### PR TITLE
Add defined type and facts to generate and export/collect keys across nodes

### PR DESCRIPTION
**Pull request by runejuhl**
This creates a fact for each found public ssh key for all users with uid above 500.
The defined type then creates an ssh_key, uses the fact to scoop up the detail and then
exports/collects it on another node.

This PR supercedes https://github.com/voxpupuli/puppet-ssh_keygen/pull/8

### CODE REVIEW

File: .fixtures.yml [hunk 1]
```diff
@@ -1,3 +1,5 @@
 fixtures:
+  symlinks:
+    ssh_keygen: "#{source_dir}"
```
Comment (alexjfisher) [inline .fixtures.yml]: This change shouldn’t be needed.

File: lib/facter/user_ssh_keys.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  Puppet::Type.type('user').instances.select do |user|
+    %w[root puppet jenkins].include?(user.name) ||
+      user.retrieve[user.property(:uid)].to_i >= 500
+  end
+end
+
+def add_facts_for_user(user)
+  sshdir = File.join(user.retrieve[user.property(:home)], '.ssh')
+
+  %w[rsa dsa].each do |keytype|
```
Comment (baurmatt) [inline lib/facter/user_ssh_keys.rb]: What about the other types defined here?
https://github.com/voxpupuli/puppet-ssh_keygen/blob/66d07750db7481da0baecc781089ed70437db295/manifests/init.pp#L43

**Comment by baurmatt** [comment]
It might me better to but all keys in a structured fact.

File: lib/facter/user_ssh_keys.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#
+# Caveats:
+#   none
+#
+# Notes:
+#   We need the pub key details for export to other nodes
+
+def login_users
+  Puppet::Type.type('user').instances.select do |user|
+    %w[root puppet jenkins].include?(user.name) ||
```
Comment (ekohl) [inline lib/facter/user_ssh_keys.rb]: Why this distinction? Isn't a more correct one based on the login shell?

File: lib/facter/user_ssh_keys.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    next unless File.file?(pubpath)
+    type, key, comment = Facter::Util::FileRead.read(pubpath).chomp.split($FS, 3)
+    fact_base = :"#{user.name}_ssh#{keytype}key"
+
+    Facter.add(fact_base) do
+      setcode { key }
+    end
+
+    Facter.add(:"#{fact_base}_comment") do
```
Comment (ekohl) [inline lib/facter/user_ssh_keys.rb]: Wouldn't a structured fact be better for this?

File: lib/facter/user_ssh_keys.rb (hunk 3 above)
Comment (runejuhl) [reply lib/facter/user_ssh_keys.rb]: Not quite sure, I just rebased the original PR #8 :smile:

I'm not sure about the best way to handle this. Maybe it should read the `SYS_UID_MIN` and `SYS_UID_MAX` values from `/etc/login.defs` and use that?

If you wanted to base it on login shells it could easily be hairy; you'd want to match `tcsh`, `bash`, `sh`, `dash`, `fish`, `ash`, `csh` and everything else. Or you could do it inverted and match `/usr/bin/nologin`, `/bin/false` etc., but there's no guarantee that you'd get it right.

File: lib/facter/user_ssh_keys.rb (hunk 4 above)
Comment (alexjfisher) [reply lib/facter/user_ssh_keys.rb]: @ekohl https://github.com/voxpupuli/puppet-ssh_keygen/pull/49#issuecomment-472027283

Comment (runejuhl) [reply lib/facter/user_ssh_keys.rb]: Indeed. I've got a WIP branch that rewrites the fact and adds specs, so expect some changes to this.

File: lib/facter/user_ssh_keys.rb (hunk 3 above)
Comment (ekohl) [reply lib/facter/user_ssh_keys.rb]: On Linux there's `shells(5)` which describes `/etc/shells`. This is commonly used to indicate valid login shells. Typically PAM uses this as a source to verify if someone can login.

Comment (runejuhl) [reply lib/facter/user_ssh_keys.rb]: Ah, perfect. Yes, that's much better. Thanks @ekohl

File: lib/facter/user_ssh_keys.rb (hunk 4 above)
Comment (anarcat) [reply lib/facter/user_ssh_keys.rb]: @runejuhl do you have that WIP branch anywhere we can review? i'd love to use this, but it would be unfortunate to work on improving a version that's already irrelevant...

**Comment by anarcat** [comment]
for what it's worth, there's at least one other implementation of such a fact I am familiar with:

https://gitlab.com/shared-puppet-modules-group/sshd/blob/master/lib/facter/ssh_keys_users.rb

you'll notice it does use a structured type, a hash where the key is the username and the value is another hash where the key is the path and the value is another hash where the key/values are the properties of each key.

in other words, it looks something like this:

```
user => {
  '/home/user/.ssh/id_rsa' => {
    type => 'rsa',
    key => 'AAAA...',
    comment => 'i like cats',
  }
}
```

Seems like a simple, straightforward structure we could adopt here.

**Comment by bschonec** [comment]
This request looks very useful.  Any chance the discussion could be revived?

**Comment by runejuhl** [comment]
@bschonec I considered reviving this following your comment, but I don't currently have a use for the feature, so I probably won't be able to find time to do this properly. You're more than welcome to take the current PR and revive it though 🙂

File: lib/facter/user_ssh_keys.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# Caveats:
+#   none
+#
+# Notes:
+#   We need the pub key details for export to other nodes
+
+def login_users
+  Puppet::Type.type('user').instances.select do |user|
+    %w[root puppet jenkins].include?(user.name) ||
+      user.retrieve[user.property(:uid)].to_i >= 500
```
Comment (ekohl) [inline lib/facter/user_ssh_keys.rb:17]: I wonder: there is a property `system` on a user. Is that properly resolved on `instances`? If so, you wouldn't need to hardcode 500. Some systems consider `>= 1000` as non-system users. See https://www.puppet.com/docs/puppet/7/type.html#user-attribute-system
